### ISSUE TITLE
ssl_server_session_cache crashes because gb_trees:delete with a non-exist key

### ISSUE THREAD

**Issue by zzydxm** [erlang/otp#5192]
**Describe the bug**
The error code is:
Supervisor: {local,ssl_upgrade_server_session_cache_sup_dist}. Context: child_terminated. Reason: {function_clause,[{gb_trees,delete_1,[{-576456945248890031,-576460752303414715},nil],[{file,"gb_trees.erl"},{line,408}]},{gb_trees,delete_1,2,[{file,"gb_trees.erl"},{line,409}]},{gb_trees,delete_1,2,[{file,"gb_trees.erl"},{line,412}]},{gb_trees,delete_1,2,[{file,"gb_trees.erl"},{line,409}]},{gb_trees,delete_1,2,[{file,"gb_trees.erl"},{line,412}]},{gb_trees,delete,2,[{file,"gb_trees.erl"},{line,404}]},{ssl_server_session_cache,handle_call,3,[{file,"ssl_server_session_cache.erl"},{line,146}]},{gen_server,try_handle_call,4,[{file,"gen_server.erl"},{line,721}]},{gen_server,handle_msg,6,[{file,"gen_server.erl"},{line,750}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}.

I think it is because when CB = ssl_server_session_cache_db, line 146 changed the Store but does not update it in gen_server state.

To fix this, we can let invalidate_session/5 to return {Store, Order} instead of just Order, and add a ssl_server_session_cache_db case for invalidate_session/5

**To Reproduce**
Start a ssl connection with session timeout 1
Wait 1 second and force it reconnect
```{_,_,_,DB,_,Order,_,_}=sys:get_state(ssl_unknown_listener).```
```gb_trees:size(DB) -  gb_trees:size(Order) > 0.```

**Expected behavior**
DB should be the same size with Order

**Affected versions**
OTP 24.0.5

**Comment by IngelaAndin** [comment]
I believe you found a bug. And that the following ought to fix it. Still working on making a good test case. 

 ```
diff --git a/lib/ssl/src/ssl_server_session_cache.erl b/lib/ssl/src/ssl_server_session_cache.erl
index 4c77fa951b..a5e4334203 100644
--- a/lib/ssl/src/ssl_server_session_cache.erl
+++ b/lib/ssl/src/ssl_server_session_cache.erl
@@ -143,8 +143,8 @@ handle_call({reuse_session, SessionId}, _From,  #state{store_cb = Cb,
                 true ->
                     {reply, Session, State0};
                 false ->
-                    Order = invalidate_session(Cb, Store0, Order0, SessionId, InId),
-                    {reply, not_reusable, State0#state{session_order = Order}}
+                    {Store, Order} = invalidate_session(Cb, Store0, Order0, SessionId, InId),
+                    {reply, not_reusable, State0#state{db = Store, session_order = Order}}
             end
     end.
 
@@ -175,7 +175,7 @@ handle_cast({register_session, #session{session_id = SessionId, time_stamp = Tim
                         true ->
                             Store = update(Cb, Store0, SessionId, Session#session{time_stamp = TimeStamp}),
                             State0#state{db = Store,
-                                         session_order = gb_trees:insert(InternalId, SessionId, Order0)};
+                                         session_order = gb_trees:insert(InternalId, SessionId, Order0)}; 
                         false ->
                             %% Throw away oldest session as it is not valid anymore
                             Store1 = delete(Cb, Store0, OldSessId),
@@ -198,7 +198,7 @@ handle_info(_, State) ->
     {noreply, State}.
 
 terminate(_, _) ->
-    ok.
+    ok. 
 
 %%%===================================================================
 %%% Internal functions
@@ -220,9 +220,9 @@ session_id(Key) ->
     Bin2 = crypto:crypto_one_time(aes_128_ecb, Key, <<Unique2:128>>, true),
     <<Bin1/binary, Bin2/binary>>.
 
-invalidate_session(Cb, Store, Order, SessionId, InternalId) ->
-    Cb:delete(Store, SessionId),
-    gb_trees:delete(InternalId, Order).
+invalidate_session(Cb, Store0, Order, SessionId, InternalId) ->
+    Store = Cb:delete(Store0, SessionId),
+    {Store, gb_trees:delete(InternalId, Order)}.
 
 init(Cb, Options) ->
     Cb:init(Options).

```

**Comment by zzydxm** [comment]
I think we should not use Store = Cb:delete(Store0, SessionId) for all here
only when Cb = ssl_server_session_cache_db we should return the result of Cb:delete, otherwise we should still return old Store, like other internal functions (update/delete)

**Comment by IngelaAndin** [comment]
Good catch it should be 

+invalidate_session(Cb, Store0, Order, SessionId, InternalId) ->
+    Store = delete(Cb, Store0, SessionId),
+    {Store, gb_trees:delete(InternalId, Order)}.

**Comment by IngelaAndin** [comment]
Just in time for OTP-24.1

**Comment by mikpe** [comment]
Head's up that we're seeing this error semi-frequently with OTP-23.3.4.9.

**Comment by IngelaAndin** [comment]
@mikpe so this bug was present in OTP-23.3.4.9  and it was fixed in OTP-24.1  I am not sure what you want to achieve with your comment!?

**Comment by mikpe** [comment]
What I meant was that the fix should be backported to maint-23 for inclusion in the next OTP-23 patch release.
